tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -2123,11 +2123,18 @@ Status ConvertConv2DHelper(OpConverterParams* params, int group,
         {{"input_sizes", true}, {"filter", true}, {"out_backprop", false}}));
     backprop_output_size = inputs.at(0);
     tensor = inputs.at(2).tensor();
-    if (!HasStaticShape(tensor->getDimensions())) {
-      // TODO(tfeher): Allow dynamic input. We need to implement padding
-      // correction for dynamic shapes in this case.
+
+    auto has_dynamic_hw_shape = [](const nvinfer1::Dims& dims) {
+      int start_idx = dims.nbDims == 4 ? 2 : 1;
+      for (int i = start_idx; i < dims.nbDims; ++i) {
+        if (dims.d[i] < 0) return true;
+      }
+      return false;
+    };
+    if (has_dynamic_hw_shape(tensor->getDimensions())) {
```

Comment:
(1) Can you add comment to explain the situations for dims.nbDims=4 and dims.nbDims!=4? I was understand the impression that nbDims is always 4 here.  For nbDims!=4, what is the data format? NHW?
(2) I don't see a test case for nbDims !=4. Am I right? Please add a test case for that.
(3) The lambda here doesn't look necessary as it is only used once.

Reply:
In the code it's very confusing, in some situations we use TRT style implicit batch dims, other times use a batch dimension is inserted, etc. I'll go back and check.

Reply:
The converter receives the input `tensor` through the input params. `tensor` is a `TensorProxyPtr`. It either has 3 dims (implicit batch) or 4 dims (explicit batch). You can see further down lines in some pre-existing lines the following code:

```
const auto tensor_dim = tensor->getDimensions();
const int c_dim_size = tensor_dim.d[params->use_implicit_batch ? 0 : 1];
```

My understanding is that in the 2D operations case, `use_implicit_batch` and `tensor_dim.nbDims != 4` are the same. `tensor_dims.nbDims==4` means `!use_implicit_batch`

Reply:
Can we write the code like this:
  if (params->use_implicit_batch) {
     return error if dims.nbDims != 3
     start_idx = 1
 } else {
    return error if dims.nbDims != 4
    start_idx = 2
 }

Reply:
Yes of course, I'll change it